linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+};
+
+template < typename T, int n0, int n1, int n2, int n3, int n4 >
+struct tensor<T, n0, n1, n2, n3, n4>
+{
+   using type = T;
+   static constexpr int ndim      = 5;
+   static constexpr int first_dim = n0;
+   MFEM_HOST_DEVICE tensor< T, n1, n2, n3, n4 >& operator[](int i) { return values[i]; }
+   MFEM_HOST_DEVICE const tensor< T, n1, n2, n3, n4 >& operator[](int i) const { return values[i]; }
+   MFEM_HOST_DEVICE tensor< T, n1, n2, n3, n4 >& operator()(int i) { return values[i]; }
+   MFEM_HOST_DEVICE const tensor< T, n1, n2, n3, n4 >& operator()(int i) const { return values[i]; }
+   MFEM_HOST_DEVICE tensor< T, n2, n3, n4 >& operator()(int i, int j) { return values[i][j]; }
+   MFEM_HOST_DEVICE const tensor< T, n2, n3, n4 >& operator()(int i,
+                                                              int j) const { return values[i][j]; }
+   MFEM_HOST_DEVICE tensor< T, n3, n4>& operator()(int i, int j, int k) { return values[i][j][k]; }
+   MFEM_HOST_DEVICE const tensor< T, n3, n4>& operator()(int i, int j,
+                                                         int k) const { return values[i][j][k]; }
+   MFEM_HOST_DEVICE tensor< T, n4 >& operator()(int i, int j, int k, int l) { return values[i][j][k][l]; }
+   MFEM_HOST_DEVICE const tensor< T, n4 >& operator()(int i, int j, int k,
+                                                      int l) const { return values[i][j][k][l]; }
+   MFEM_HOST_DEVICE T& operator()(int i, int j, int k, int l, int m) { return values[i][j][k][l][m]; }
+   MFEM_HOST_DEVICE const T& operator()(int i, int j, int k, int l, int m) const { return values[i][j][k][l][m]; }
+   tensor < T, n1, n2, n3, n4 > values[n0];
+};
+
+/**
+ * @brief A sentinel struct for eliding no-op tensor operations
+ */
+struct zero
+{
+   /** @brief `zero` is implicitly convertible to double with value 0.0 */
+   MFEM_HOST_DEVICE operator double() { return 0.0; }
+
+   /** @brief `zero` is implicitly convertible to a tensor of any shape */
+   template <typename T, int... n>
+   MFEM_HOST_DEVICE operator tensor<T, n...>()
+   {
+      return tensor<T, n...> {};
+   }
+
+   /** @brief `zero` can be accessed like a multidimensional array */
+   template <typename... T>
+   MFEM_HOST_DEVICE zero operator()(T...)
+   {
+      return zero{};
+   }
+
+   /** @brief anything assigned to `zero` does not change its value and returns `zero` */
+   template <typename T>
+   MFEM_HOST_DEVICE zero operator=(T)
+   {
+      return zero{};
+   }
+};
+
+/** @brief checks if a type is `zero` */
+template <typename T>
+struct is_zero : std::false_type
+{
+};
+
+/** @overload */
+template <>
+struct is_zero<zero> : std::true_type
+{
+};
+
+/** @brief the sum of two `zero`s is `zero` */
+MFEM_HOST_DEVICE constexpr zero operator+(zero, zero) { return zero{}; }
+
+/** @brief the sum of `zero` with something non-`zero` just returns the other value */
+template <typename T>
+MFEM_HOST_DEVICE constexpr T operator+(zero, T other)
```

Comment:
I think it is better to pass `T` by reference here. Same applies to other similar arguments below.

Reply:
when compiling with -O1 or greater, it doesn't make any difference: https://godbolt.org/z/Tno3MeEe5

interestingly, when compiling this without optimization, pass-by-value emits fewer instructions than by-reference, so it should be slightly faster (although this is unimportant)

Reply:
oops, I misread the linked code snippet: the operator+ does need the value (or a reference to it), unlike operator*

the generated assembly for a few different implementations of operator+ are given here: https://godbolt.org/z/znPYG4Ybf

Reply:
My thinking was for the case when `T` is not just a simple type like `double` but a `tensor` or something like that.